### PR TITLE
Added UpdateModel, publishModel,and unpublishModel functionality + tests

### PR DESCRIPTION
**Pull request by ifielker**
Added UpdateModel, publishModel,and unpublishModel functionality + tests

### CODE REVIEW

File: src/machine-learning/machine-learning-api-client.ts [hunk 1]
```diff
@@ -31,6 +31,20 @@ export interface StatusErrorResponse {
     readonly message: string;
 }
 
+/**
+ * A Firebase ML Model input object
+ */
+export class ModelOptions {
```
Comment (hiranya911) [inline src/machine-learning/machine-learning-api-client.ts]: interface

File: src/machine-learning/machine-learning-api-client.ts [hunk 2]
```diff
@@ -31,6 +31,20 @@ export interface StatusErrorResponse {
...
+  public displayName?: string;
+  public tags?: string[];
+
+  public tfliteModel?: { gcsTfliteUri: string; };
+}
+
+export class ModelUpdateOptions extends ModelOptions {
```
Comment (hiranya911) [inline src/machine-learning/machine-learning-api-client.ts]: interface

File: src/machine-learning/machine-learning.ts [hunk 3]
```diff
@@ -288,22 +304,26 @@ export interface TFLiteModel {
...
- * A Firebase ML Model input object
- */
-export class ModelOptions {
-  public displayName?: string;
-  public tags?: string[];
 
-  public tfliteModel?: { gcsTfliteUri: string; };
-}
 
+function getMaskFromOptions(options: ModelOptions): string[] {
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts]: Can we use the existing `generateUpdateMask()` function in `utils/index.ts`?

File: src/machine-learning/machine-learning.ts [hunk 4]
```diff
@@ -109,8 +110,13 @@ export class MachineLearning implements FirebaseServiceInterface {
    * @return {Promise<Model>} A Promise fulfilled with the updated model.
    */
   public updateModel(modelId: string, model: ModelOptions): Promise<Model> {
-    throw new Error('NotImplemented');
-  }
+     const updateMask = getMaskFromOptions(model);
+     return this.maybeSignUrl(model, true)
+      .then((modelContent) => this.client.updateModel(modelId, modelContent, updateMask))
+      .then((operation) => {
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts]: Let's be consistent. Either use the one-liner format or use the block format.

```
.then((x) => doStuff(x))
```

or 

```
.then((x) => {
  return doStuff(x);
})
```

I personally prefer the latter. But let's use one format throughout the file.

File: src/machine-learning/machine-learning.ts [hunk 5]
```diff
@@ -131,7 +142,12 @@ export class MachineLearning implements FirebaseServiceInterface {
    * @return {Promise<Model>} A Promise fulfilled with the unpublished model.
    */
   public unpublishModel(modelId: string): Promise<Model> {
-    throw new Error('NotImplemented');
+    const updateMask = ['state.published'];
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts]: This is almost identical to `publishModel` except for the value of `published` property. May be refactor into a helper function?

```
private setPublishStatus(modelId: string, published: boolean): Promise<Model>
```

File: src/machine-learning/machine-learning.ts [hunk 6]
```diff
@@ -171,23 +187,23 @@ export class MachineLearning implements FirebaseServiceInterface {
     return this.client.deleteModel(modelId);
   }
 
-  private convertOptionstoContent(options: ModelOptions, forUpload?: boolean): Promise<ModelContent> {
-    const modelContent = deepCopy(options);
+  private maybeSignUrl(options: ModelOptions, forUpload?: boolean): Promise<ModelOptions> {
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts]: `maybe` sounds like it's non-deterministic. Perhaps just `signUrlIfPresent()`?

Also is `forUpload` ever `false`?

File: src/machine-learning/machine-learning.ts [hunk 7]
```diff
@@ -288,22 +304,26 @@ export interface TFLiteModel {
...
+function getMaskFromOptions(options: ModelOptions): string[] {
+  const mask: string[] = [];
+  if (options.displayName) {
+    mask.push('display_name');
+  }
+  if (options.tags) {
+    mask.push('tags');
+  }
+  if (options.tfliteModel) {
+    mask.push('tflite_model.gcs_tflite_uri');
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts]: You have `tfliteModel` in the body but `tflite_model` in the mask here. Same for `gcs_tflite_uri`.

File: test/integration/machine-learning.spec.ts [hunk 8]
```diff
@@ -127,6 +127,177 @@ describe('admin.machineLearning', () => {
     });
   });
 
+  describe('updateModel()', () => {
+    const UPDATE_NAME: admin.machineLearning.ModelOptions = {
+      displayName: 'update-model-new-name',
+    };
```
Comment (hiranya911) [inline test/integration/machine-learning.spec.ts]: Add newline

File: test/integration/machine-learning.spec.ts [hunk 9]
```diff
@@ -127,6 +127,177 @@ describe('admin.machineLearning', () => {
...
+          const modelOptions: admin.machineLearning.ModelOptions = {
+            displayName: DISPLAY_NAME,
+          };
+          return admin.machineLearning().updateModel(model.modelId, modelOptions)
+            .then((updatedModel) => {
+              verifyModel(updatedModel, modelOptions);
+            });
+        });
+    });
+
```
Comment (hiranya911) [inline test/integration/machine-learning.spec.ts]: Nit: Remove extra empty line

File: src/machine-learning/machine-learning-api-client.ts (hunk 1 above)
Comment (ifielker) [reply src/machine-learning/machine-learning-api-client.ts]: done

File: src/machine-learning/machine-learning-api-client.ts (hunk 2 above)
Comment (ifielker) [reply src/machine-learning/machine-learning-api-client.ts]: done

File: src/machine-learning/machine-learning.ts (hunk 3 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts]: No, utils.generateUpdateMask gives you camelCase, we need snake_case. The mask is required to matched the proto field name and protos are in snake_case. We could change the ModelOptions definition to be in snake_case, but I think that would not be a good experience for Node.js developers who are used to camelCase field names.

File: src/machine-learning/machine-learning.ts (hunk 4 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts]: Oops, sorry I had some logging in there for debugging and forgot to put it back.

File: src/machine-learning/machine-learning.ts (hunk 5 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts]: done

File: src/machine-learning/machine-learning.ts (hunk 6 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts]: done

File: src/machine-learning/machine-learning.ts (hunk 7 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts]: Yes, this is intentional. Node.js field names are in camelCase, but mask must be snake_case to match the proto.

File: test/integration/machine-learning.spec.ts (hunk 8 above)
Comment (ifielker) [reply test/integration/machine-learning.spec.ts]: done

File: test/integration/machine-learning.spec.ts (hunk 9 above)
Comment (ifielker) [reply test/integration/machine-learning.spec.ts]: done

File: src/machine-learning/machine-learning.ts [hunk 10]
```diff
@@ -224,9 +216,7 @@ export class MachineLearning implements FirebaseServiceInterface {
     return blob.getSignedUrl({
       action: 'read',
       expires: Date.now() + URL_VALID_DURATION,
-    }).then((x) => {
-      return x[0];
-    });
+    }).then((x) => x[0]);
```
Comment (hiranya911) [inline src/machine-learning/machine-learning.ts]: Nit: Replace x with something more descriptive. `signResult` or similar.

File: test/unit/machine-learning/machine-learning-api-client.spec.ts [hunk 11]
```diff
@@ -189,6 +189,124 @@ describe('MachineLearningApiClient', () => {
...
+      const stub = sinon
+        .stub(HttpClient.prototype, 'send')
+        .rejects(utils.errorFrom(ERROR_RESPONSE, 404));
+      stubs.push(stub);
+      const expected = new FirebaseMachineLearningError('not-found', 'Requested entity not found');
+      return apiClient.updateModel(MODEL_ID, NAME_ONLY_CONTENT, NAME_ONLY_MASK)
+        .should.eventually.be.rejected.and.deep.equal(expected);
+    });
+
+    it('should resolve with the created resource on success', () => {
```
Comment (hiranya911) [inline test/unit/machine-learning/machine-learning-api-client.spec.ts]: s/created/updated

File: test/unit/machine-learning/machine-learning-api-client.spec.ts [hunk 12]
```diff
@@ -189,6 +189,124 @@ describe('MachineLearningApiClient', () => {
...
+      const stub = sinon
+        .stub(HttpClient.prototype, 'send')
+        .resolves(utils.responseFrom(OPERATION_SUCCESS_RESPONSE));
+      stubs.push(stub);
+      return apiClient.updateModel(MODEL_ID, NAME_ONLY_CONTENT, NAME_ONLY_MASK)
```
Comment (hiranya911) [inline test/unit/machine-learning/machine-learning-api-client.spec.ts]: Do an assertion on the URL and the body (Make sure the correct payload and mask is being passed to the backend)

File: src/machine-learning/machine-learning.ts (hunk 10 above)
Comment (ifielker) [reply src/machine-learning/machine-learning.ts]: done

File: test/unit/machine-learning/machine-learning-api-client.spec.ts (hunk 11 above)
Comment (ifielker) [reply test/unit/machine-learning/machine-learning-api-client.spec.ts]: done

File: test/unit/machine-learning/machine-learning-api-client.spec.ts (hunk 12 above)
Comment (ifielker) [reply test/unit/machine-learning/machine-learning-api-client.spec.ts]: done

**Pull request merged**